{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "04b4a77f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "import os\n",
    "import re\n",
    "\n",
    "import sqlite3\n",
    "import pandas as pd\n",
    "import plotly.express as px\n",
    "from plotly.subplots import make_subplots\n",
    "import plotly.graph_objects as go\n",
    "import igraph\n",
    "from igraph import Graph, EdgeSeq\n",
    "from jinja2 import Template\n",
    "\n",
    "NTER_CMAP = {\n",
    "        \"Z-type\":\"#7ad5a3\",\n",
    "        \"X-type\":\"#322f26\",\n",
    "        \"CoBaHMA-type\":\"#0babc1\",\n",
    "        \"Y-type\":\"#b980d1\",\n",
    "        \"Unknown-type\":\"#717171\"\n",
    "}\n",
    "\n",
    "def define_ccya_genotype(x):\n",
    "    if x.Accession.startswith(\"GCA\"):\n",
    "        prefix=\"GCA\" \n",
    "    elif x.Accession.startswith(\"GCF\"):\n",
    "        prefix = \"GCF\"\n",
    "    else:\n",
    "        return None\n",
    "    \n",
    "    if x.flag == \"Calcyanin with known N-ter\" or x.flag == \"Calcyanin with new N-ter\":\n",
    "        return prefix + \"+\"\n",
    "    return prefix+\"-\"\n",
    "\n",
    "def count_genotypes(df , group):\n",
    "    datas = {\"ccyA+\":0,\"ccyA-\":0}\n",
    "    if group in df.columns:\n",
    "        df = df.reset_index().set_index(group)\n",
    "        for k in list(set(df.index)):\n",
    "            #datas[k]=\"ccyA-\"\n",
    "            loc = df.loc[k]\n",
    "            if isinstance(loc,pd.Series):\n",
    "                loc = pd.DataFrame(loc).T\n",
    "            if \"ccyA+\" in loc[\"(ccyA)\"].values:                \n",
    "                datas[\"ccyA+\"]+=1\n",
    "                continue\n",
    "            datas[\"ccyA-\"]+=1\n",
    "        return datas\n",
    "    else:\n",
    "        raise KeyError(group)\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "4da35165",
   "metadata": {},
   "outputs": [],
   "source": [
    "db = \"../../../tests_dir/micro_datasets/wpcalf.db\"\n",
    "dbname = os.path.basename(db).split(\".\")[0]\n",
    "cnx = sqlite3.connect(db)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "5b3b781a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_genome_pie_chart(cnx):\n",
    "    summary_df = pd.read_sql_query(\"\"\"\n",
    "        SELECT * FROM genomes as g JOIN \n",
    "        harley as h on h.Accession=g.Accession JOIN \n",
    "        summary as s on G.Accession = s.sequence_src\"\"\", \n",
    "                                   cnx , index_col=\"Accession\").reset_index()\n",
    "    summary_df.Accession = summary_df.apply(lambda x : x.Accession[0],axis=1)\n",
    "    summary_df[\"uid\"] = summary_df.apply(lambda x : x.Accession.split(\"_\")[-1].split(\".\")[0], axis=1 )\n",
    "    summary_df[\"uidv\"] = summary_df.apply(lambda x : x.Accession.split(\"_\")[-1], axis=1 )\n",
    "    summary_df[\"(DBccyA)\"] = summary_df.apply(lambda x : define_ccya_genotype(x), axis=1 )\n",
    "    summary_df[\"(ccyA)\"] = summary_df.apply(lambda x : \"ccyA+\" if x.flag in [\"Calcyanin with known N-ter\",\"Calcyanin with new N-ter\"] else \"ccyA-\", axis=1 )\n",
    "    summary_df = summary_df[[\"Organism\", \"uid\", \"uidv\", \"Accession\", \"Assembly name\" , \"Date\" , \"(ccyA)\" , \"sequence_accession\", \"flag\", \"nter\"]]\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "    # We reformat the dictionnary and make a plotly-friendly dataframe.\n",
    "    metrics_d = []\n",
    "    for g in [\"Organism\", \"uid\" , \"uidv\" , \"Accession\"] :\n",
    "        d = count_genotypes(summary_df , group = g)      \n",
    "        total = len(list(summary_df[g].unique()))\n",
    "        metrics_d.append(\n",
    "            (\"{} [{}]\".format(g,total) ,\"ccyA+\",d[\"ccyA+\"])\n",
    "        )\n",
    "        metrics_d.append(\n",
    "            (\"{} [{}]\".format(g,total),\"ccyA-\",d[\"ccyA-\"])\n",
    "        )\n",
    "\n",
    "\n",
    "    # We make a figure with multiple pie-chart [strain, GCA-GCF, GCA, GCF].   \n",
    "    metrics_df = pd.DataFrame(metrics_d)\n",
    "    metrics_df.columns = [\"RedLevel\",\"genotype\",\"count\"]\n",
    "    fig = px.pie(metrics_df,values=\"count\",names=\"genotype\",facet_col=\"RedLevel\",color=\"genotype\",\n",
    "            color_discrete_map={\"ccyA-\":\"#A61C3C\",\"ccyA+\":\"#439775\",\"ccyA~\":\"orange\"})\n",
    "    return fig\n",
    "\n",
    "#make_genome_pie_chart(cnx)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "84502de0",
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_decision_tree_chart():\n",
    "    # DECISION TREE CHART\n",
    "        nr_vertices = 11\n",
    "        v_label = list(map(str, range(nr_vertices)))\n",
    "        G = Graph.Tree(nr_vertices, 2) # 2 stands for children number\n",
    "        lay = G.layout_reingold_tilford(mode=\"in\", root=[0])\n",
    "        position = {k: lay[k] for k in range(nr_vertices)}\n",
    "        Y = [lay[k][1] for k in range(nr_vertices)]\n",
    "        M = max(Y)\n",
    "        es = EdgeSeq(G) # sequence of edges\n",
    "        E = [e.tuple for e in G.es] # list of edges\n",
    "        L = len(position)\n",
    "        Xn = [position[k][0] for k in range(L)]\n",
    "        Yn = [2*M-position[k][1] for k in range(L)]\n",
    "        Xe = []\n",
    "        Ye = []\n",
    "        Ec = {}\n",
    "        color = \"#c6587e\"\n",
    "        txt = \"N\"\n",
    "        for edge in E:\n",
    "            Xe=[position[edge[0]][0],position[edge[1]][0], None]\n",
    "            Ye=[2*M-position[edge[0]][1],2*M-position[edge[1]][1], None]\n",
    "            Ec[edge] = [Xe,Ye,color,txt]\n",
    "            # we alternate the vertice color for positive and negative answer\n",
    "            if color == \"#48d38b\":\n",
    "                color = \"#c6587e\"\n",
    "                txt = \"N\"\n",
    "            else:\n",
    "                color = \"#48d38b\"\n",
    "                txt = \"Y\"\n",
    "        # Nodes values\n",
    "        node_txt = {\n",
    "            0:\"[Sequence has a significative hit against the GlyX3]<br>Does the sequence have three glycine zipper in the right order (G1|G2|G3) ?\",\n",
    "            1:\"Does the sequence have<br>at least a G1 and G3 in this order ?\",\n",
    "            2:\"Does the sequence have<br>a Known N-ter ?\",\n",
    "            3:\"Does the sequence have<br>Known N-ter ?\",\n",
    "            4:\"Does the sequence have<br>a N-ter of type Y ?\",\n",
    "            5:\"Calcyanin with<br>new N-ter\",\n",
    "            6:\"Calcyanin with<br>known N-ter\",\n",
    "            7:\"Atypical gly region<br>with new N-ter\",\n",
    "            8:\"Atypical gly region<br>with known N-ter\",\n",
    "            9:\"Atypical gly region<br>with new N-ter\",#\"Calcyanin with<br>new N-ter\",\n",
    "            10:\"Calcyanin with<br>known N-ter\",\n",
    "        }  \n",
    "\n",
    "        labels = [\"<b>\"+txt+\"</b>\" if _ in [9,10,5,6] else txt for _,txt in node_txt.items()]\n",
    "        decision_tree = go.Figure()\n",
    "        for e,E in Ec.items():\n",
    "            decision_tree.add_trace(go.Scatter(x=E[0],\n",
    "                            y=E[1],\n",
    "                        mode='lines',\n",
    "                        line=dict(color=E[2], width=4),\n",
    "                        opacity=0.5,\n",
    "                        text = E[3],\n",
    "                        hoverinfo=\"text\",\n",
    "        #                   hoverinfo='none'\n",
    "                        ))\n",
    "\n",
    "\n",
    "        decision_tree.add_trace(go.Scatter(x=Xn,\n",
    "                        y=Yn,\n",
    "                        mode='markers+text',\n",
    "                        name='bla',\n",
    "                        marker=dict(symbol='diamond',\n",
    "                                        size=30,\n",
    "                                        color='white',    #'#DB4551',\n",
    "                                        #opacity=0.5,\n",
    "                                        line=dict(color='rgb(50,50,50)', width=1)\n",
    "                                        ),\n",
    "                        text=labels,\n",
    "                        textposition='top center',\n",
    "                        hoverinfo='text',\n",
    "                        opacity=1\n",
    "                        ))\n",
    "        decision_tree.update_xaxes(visible=False)\n",
    "        decision_tree.update_yaxes(visible=False)\n",
    "        decision_tree.update_layout({\n",
    "        'margin':dict(l=40, r=40, b=85, t=100),\n",
    "        'showlegend':False,\n",
    "        'plot_bgcolor': 'rgba(0, 0, 0, 0)',\n",
    "        'paper_bgcolor': 'rgba(0, 0, 0, 0)',\n",
    "        'title': \"Calcyanin classification decision tree.\",\n",
    "        'height':750,\n",
    "        })\n",
    "        return decision_tree\n",
    "#make_decision_tree_chart()    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2e379dc5",
   "metadata": {},
   "source": [
    "# Number of cyanobacteria over time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "2219a39c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def count_genome_by_date(df , group , label):\n",
    "    df.sort_values(\"Date\",inplace=True)\n",
    "    if group in df.columns:\n",
    "        df = df.reset_index().set_index(group)\n",
    "        df = df[~df.index.duplicated(keep=\"first\")]\n",
    "        df.reset_index(inplace=True)\n",
    "        df = pd.DataFrame(df.groupby(\"Date\").count()[group])\n",
    "        df.columns=[\"count_by_date\"]\n",
    "        df[\"total\"] = df[\"count_by_date\"].cumsum(axis = 0)\n",
    "        df[\"label\"] = label\n",
    "        return df\n",
    "    else:\n",
    "        raise KeyError(group)\n",
    "\n",
    "def make_genome_over_time_chart(cnx):\n",
    "    summary_df = pd.read_sql_query(\"\"\"\n",
    "        SELECT * FROM genomes as g JOIN \n",
    "        harley as h on h.Accession=g.Accession JOIN \n",
    "        summary as s on G.Accession = s.sequence_src\"\"\", \n",
    "                                   cnx , index_col=\"Accession\").reset_index()\n",
    "    summary_df.Accession = summary_df.apply(lambda x : x.Accession[0],axis=1)\n",
    "    summary_df[\"uid\"] = summary_df.apply(lambda x : x.Accession.split(\"_\")[-1].split(\".\")[0], axis=1 )\n",
    "    summary_df[\"uidv\"] = summary_df.apply(lambda x : x.Accession.split(\"_\")[-1], axis=1 )\n",
    "    summary_df[\"(DBccyA)\"] = summary_df.apply(lambda x : define_ccya_genotype(x), axis=1 )\n",
    "    summary_df[\"(ccyA)\"] = summary_df.apply(lambda x : \"ccyA+\" if x.flag in [\"Calcyanin with known N-ter\",\"Calcyanin with new N-ter\"] else \"ccyA-\", axis=1 )\n",
    "    summary_df = summary_df[[\"Organism\", \"uid\", \"uidv\", \"Accession\", \"Assembly name\" , \"Date\" , \"(ccyA)\" , \"sequence_accession\", \"flag\", \"nter\"]]\n",
    "\n",
    "\n",
    "    assembly_count_df = pd.concat(\n",
    "        [count_genome_by_date(summary_df,\"Organism\",\"Strain<br>[i.e Microcystis aeruginosa PCC 9443]\"),\n",
    "         count_genome_by_date(summary_df,\"uid\",\"Assembly<br>[i.e XXX_<assembly>.N]\"),\n",
    "         count_genome_by_date(summary_df,\"uidv\",\"Version<br>[i.e XXX_<assembly>.<version>]\"),     \n",
    "         count_genome_by_date(summary_df,\"Accession\",\"Entry<br>[i.e <ncbi db>_<assembly>.<version>]\"),\n",
    "        ])\n",
    "\n",
    "\n",
    "    genome_over_time = px.line(assembly_count_df.reset_index().sort_values([\"Date\",\"label\"]),\n",
    "                               hover_data=[\"count_by_date\"],\n",
    "                               x=\"Date\", y=\"total\",color=\"label\" ,markers=True)   \n",
    "    genome_over_time.update_layout(\n",
    "        title=\"Number of entry over time\",\n",
    "        yaxis_title=\"No entry\",\n",
    "        xaxis_title=\"Date\",\n",
    "        legend_title=\"Level of redundancy\",\n",
    "        paper_bgcolor='rgba(0,0,0,0)',\n",
    "        plot_bgcolor='rgba(0,0,0,0)',\n",
    "        xaxis=dict(showgrid=False),\n",
    "        yaxis=dict(gridcolor=\"rgba(0,0,0,0.1)\")\n",
    "    )\n",
    "    return genome_over_time\n",
    "\n",
    "#make_genome_over_time_chart(cnx)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b7da5eea",
   "metadata": {},
   "source": [
    "# Number of sequences over time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "bb4a44f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def count_seq_by_date(df , group , subset = [] , label = None):\n",
    "    df.sort_values(\"Date\",inplace=True)\n",
    "    if group in df.columns:\n",
    "        df = df.reset_index().set_index(group)\n",
    "        df = df[~df.index.duplicated(keep=\"first\")]\n",
    "\n",
    "        df.reset_index(inplace=True)\n",
    "        df.fillna(\"NA\",inplace=True)\n",
    "        grouping_variables = [\"Date\"]+subset\n",
    "\n",
    "        df = pd.DataFrame(df.groupby(grouping_variables).count()[group])\n",
    "        df.columns=[\"count_by_date\"]\n",
    "        #df[\"total\"] = df[\"count_by_date\"].cumsum(axis = 0)\n",
    "        #df[\"label\"] = label if label else group\n",
    "        return df\n",
    "    else:\n",
    "        raise KeyError(group)\n",
    "\n",
    "def make_sequence_over_time_chart(cnx):\n",
    "    summary_df = pd.read_sql_query(\"\"\"\n",
    "        SELECT * FROM genomes as g JOIN \n",
    "        harley as h on h.Accession=g.Accession JOIN \n",
    "        summary as s on G.Accession = s.sequence_src\"\"\", \n",
    "                                   cnx , index_col=\"Accession\").reset_index()\n",
    "    summary_df.Accession = summary_df.apply(lambda x : x.Accession[0],axis=1)\n",
    "\n",
    "    assembly_count_df = count_seq_by_date(summary_df,\"sequence_accession\", [\"flag\",\"nter\"])\n",
    "\n",
    "\n",
    "    datas = []\n",
    "    for flag,nter in assembly_count_df.reset_index(\"Date\").index.unique():\n",
    "        tmp = assembly_count_df.reset_index()\n",
    "        tmp = tmp[(tmp.flag == flag) & (tmp.nter == nter)]\n",
    "        tmp[\"total\"] = tmp.count_by_date.cumsum(axis = 0)\n",
    "        datas.append(tmp)\n",
    "    assembly_count_df = pd.concat(datas)\n",
    "\n",
    "\n",
    "    fig = px.line(assembly_count_df.reset_index(),x=\"Date\", y=\"total\",\n",
    "                               color_discrete_map=NTER_CMAP,\n",
    "                               hover_data=[\"count_by_date\"],color=\"nter\",line_dash=\"flag\" ,markers=True)   \n",
    "    fig.update_layout(\n",
    "        title=\"Number of calcyanin by N-ter type and flag over time\",\n",
    "        yaxis_title=\"No sequences\",\n",
    "        xaxis_title=\"Date\",\n",
    "        legend_title=\"Calcyanin N-ter type and flag\",\n",
    "        paper_bgcolor='rgba(0,0,0,0)',\n",
    "        plot_bgcolor='rgba(0,0,0,0)',\n",
    "        xaxis=dict(showgrid=False),\n",
    "        yaxis=dict(gridcolor=\"rgba(0,0,0,0.1)\")\n",
    "    )\n",
    "    return fig\n",
    "\n",
    "#make_sequence_over_time_chart(cnx)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8108b59c",
   "metadata": {},
   "source": [
    "# Modular organization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "50ee93ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "DOM_CMAP = {\n",
    "        \"Gly1\":\"#ecec13\",   \n",
    "        \"Gly2\":\"#ec9d13\",\n",
    "        \"Gly3\":\"#e25d3f\",\n",
    "        \"GlyX3\":\"#d9d5d4\"\n",
    "    }\n",
    "DOM_CMAP.update(NTER_CMAP)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "587c1835",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "def sequence_modular_orga(rid,record,y,**kwargs):\n",
    "        \"\"\"\n",
    "        rid : Sequence accession\n",
    "        record : \n",
    "        \"\"\"\n",
    "        traces = []\n",
    "        segments = {}\n",
    "        segment_id = 0\n",
    "        for _,f in record[\"features\"].items():\n",
    "            domid = f[\"feature_id\"]     \n",
    "            if domid == \"N-ter\":\n",
    "                domid = f[\"feature_src\"].split(\"|\")[0]\n",
    "\n",
    "            x = list(range(f[\"feature_start\"],f[\"feature_end\"],20))\n",
    "            hover = [domid for _ in x]\n",
    "            segments[segment_id]= (x,hover)\n",
    "            segment_id+=1\n",
    "\n",
    "\n",
    "        # ADD FULL SEQ TRACES\n",
    "        line = dict(color='black', width=1, dash='dash')\n",
    "        xlen = [0,len(record['sequence'])] #list(range(0,len(record['sequence']),len(record['sequence'])))\n",
    "        traces.append(go.Scatter(x=xlen, y=[y for i in range(0,len(xlen))],\n",
    "                            mode='lines',\n",
    "                            name=\"full-seq\",\n",
    "                            hovertext=rid,\n",
    "                            line=line,#showlegend=False,                                             \n",
    "                            legendgroup=\"full-seq\",\n",
    "                            legendgrouptitle_text=\"full-seq\",\n",
    "                            opacity = 0.4,\n",
    "                            #fill=\"none\",\n",
    "                            #fillcolor=None,\n",
    "                            #name=\"second legend group\",                                                            \n",
    "                            )\n",
    "            ) \n",
    "                \n",
    "        for _,segm in segments.items():\n",
    "                if segm:\n",
    "                    if segm[0][0] and segm[0][1] :\n",
    "                        sx,sh = segm\n",
    "                        sy = [y for i in range(0,len(sx))]\n",
    "                        dom = sh[0]\n",
    "                        if dom in DOM_CMAP:                \n",
    "                            line = dict(color=DOM_CMAP[dom],width=3)\n",
    "                        else:\n",
    "                            #print(dom)\n",
    "                            line = dict(color='black', width=1)\n",
    "\n",
    "                        neighbor = \"NA\"\n",
    "                        if \"nter_neighbor\" in record:\n",
    "                            neighbor = record[\"nter_neighbor\"]\n",
    "                        htxt = \"- {}<br>- {}<br>- {} [neighbor]<br>- {}<br>- {}<br>\".format(rid,dom , neighbor , record[\"flag\"] ,\"From {} to {}\".format(min(sx),max(sx)))\n",
    "                        for i,j in kwargs.items():\n",
    "                            htxt+=\"- {}: {}<br>\".format(i,j)\n",
    "\n",
    "                        traces.append(go.Scatter(\n",
    "                            x=sx, y=sy,\n",
    "                            mode='lines',\n",
    "                            name=dom,\n",
    "                            line=line,#showlegend=False,                                             \n",
    "                            legendgroup=dom,\n",
    "                            legendgrouptitle_text=dom,\n",
    "                            hovertext=htxt,\n",
    "                            #opacity = 0.4,\n",
    "                            #fill=\"none\",\n",
    "                            #fillcolor=None,\n",
    "                            #name=\"second legend group\",                                                            \n",
    "                            )\n",
    "                        )    \n",
    "        return traces\n",
    "\n",
    "\n",
    "def make_modorg_chart(cnx):\n",
    "    # Plotting modular organization\n",
    "    # First we load the feature 'table'\n",
    "    features_df = pd.read_sql_query(\"SELECT * FROM features as f JOIN summary as s on s.sequence_accession=f.sequence_id\", cnx ).set_index(\"sequence_id\")\n",
    "    features_df[\"e-value\"] = features_df.apply(lambda x : '{:0.2e}'.format(x[\"e-value\"]),axis=1)\n",
    "    features_dict = {}\n",
    "    # dictionnary with sequence identifiers as keys and features as values\n",
    "    for _ , sdf in features_df.groupby(\"sequence_id\"):\n",
    "        features_dict[_]=sdf.reset_index().T.to_dict()\n",
    "    features_dict\n",
    "    # dictionnary with sequence identifiers as keys and sequence-related information + features as values\n",
    "    sequence_datas = {}\n",
    "\n",
    "    sequence_df = pd.read_sql_query(\"SELECT * FROM summary as s JOIN genomes as g on g.Accession = s.sequence_src\", cnx ).set_index(\"sequence_accession\")\n",
    "    for _ , row in sequence_df.iterrows():\n",
    "        sequence_datas[_] = row.to_dict()\n",
    "        sequence_datas[_][\"features\"]=features_dict[_]  \n",
    "\n",
    "    # For each type of N-ter we make a modular organization chart.    \n",
    "    oms_plots = {}\n",
    "    for nter_type in list(set(sequence_df.nter)):\n",
    "        if not nter_type:\n",
    "            nter_type = \"Unknown-type\"\n",
    "        fig = go.Figure()\n",
    "        y = 1\n",
    "        cpt = 0\n",
    "        for rid , record in sequence_datas.items():\n",
    "            if not record[\"nter\"]:\n",
    "                record[\"nter\"] = \"Unknown-type\"\n",
    "            if record[\"nter\"] == nter_type:\n",
    "                cpt+=1\n",
    "                traces = sequence_modular_orga(rid,\n",
    "                                    record,\n",
    "                                    y,\n",
    "                                    Organism_Name=record[\"Organism\"],\n",
    "                                    Assembly=record[\"sequence_src\"]\n",
    "                                )\n",
    "                if traces:\n",
    "                    for t in traces:\n",
    "                        fig.add_trace(t)\n",
    "                    y+=1\n",
    "                cpt+=1\n",
    "                leg = []\n",
    "                for trace in fig['data']:\n",
    "                    leg.append\n",
    "                    if trace['name'] not in leg:\n",
    "                        leg.append(trace['name'])\n",
    "                    else:\n",
    "                        trace['showlegend'] = False\n",
    "            fig.update_layout(\n",
    "                autosize=False,\n",
    "                width=1000,\n",
    "                height=10*cpt if cpt > 100 else 700,\n",
    "                margin=dict(\n",
    "                    l=50,\n",
    "                    r=50,\n",
    "                    b=100,\n",
    "                    t=100,\n",
    "                    pad=4\n",
    "                ),\n",
    "                paper_bgcolor='rgba(0,0,0,0)',\n",
    "                plot_bgcolor='rgba(0,0,0,0)',\n",
    "                yaxis=dict(visible=False),\n",
    "                xaxis=dict(gridcolor=\"rgba(0,0,0,0.1)\"),\n",
    "                xaxis_title=\"Sequence length (aa)\",\n",
    "                title=\"Modular organization [{}]\".format(nter_type)\n",
    "            )   \n",
    "            oms = fig #.to_html().split(\"<body>\")[1].split(\"</body>\")[0]      \n",
    "            oms_plots[nter_type] = oms\n",
    "    return oms_plots\n",
    "\n",
    "#make_modorg_chart(cnx)[\"Z-type\"].show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "c00eedc5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_sunburst(cnx):\n",
    "    sequence_df = pd.read_sql_query(\"\"\"\n",
    "        SELECT * FROM summary as s JOIN \n",
    "        genomes as g on g.Accession = s.sequence_src JOIN\n",
    "        harley as h on h.Accession = g.Accession\n",
    "\n",
    "        \"\"\",cnx).set_index(\"sequence_accession\")\n",
    "    sequence_df.nter.fillna(\"Unknown-type\",inplace=True)\n",
    "    sequence_df = sequence_df.groupby(\n",
    "        [\"Date\",\"flag\",\"nter\",\"cter\",\"iteration\"]).count().reset_index() \n",
    "    sequence_df = sequence_df[[\"Date\",\"flag\",\"nter\",\"cter\",\"iteration\",\"sequence_src\"]]\n",
    "    sequence_df.columns = [\"Date\",\"flag\",\"nter\",\"cter\",\"iteration\",\"No of sequences\"]\n",
    "    # We make a sunburst chart : \n",
    "    sunburst = px.sunburst(sequence_df, path=['nter', 'flag', 'cter', \"Date\", 'iteration'], \n",
    "                           values='No of sequences', \n",
    "                           color = \"nter\",\n",
    "                           color_discrete_map=NTER_CMAP)\n",
    "    sunburst.update_layout(title=\"No of Calcyanin.\")\n",
    "    return sunburst\n",
    "\n",
    "def make_calcyanin_treemap(cnx):\n",
    "    sequence_df = pd.read_sql_query(\"\"\"\n",
    "        SELECT * FROM summary as s JOIN \n",
    "        genomes as g on g.Accession = s.sequence_src JOIN\n",
    "        harley as h on h.Accession = g.Accession\n",
    "\n",
    "        \"\"\",cnx).set_index(\"sequence_accession\")\n",
    "    sequence_df.nter.fillna(\"Unknown-type\",inplace=True)\n",
    "    sequence_df = sequence_df[[\"Date\",\"flag\",\"nter\",\"cter\",\"iteration\",\"sequence_src\"]]\n",
    "    sequence_df.columns = [\"Date\",\"flag\",\"nter\",\"cter\",\"iteration\",\"No of sequences\"]\n",
    "    sequence_df = sequence_df.groupby(\n",
    "        [\"Date\",\"flag\",\"nter\",\"cter\",\"iteration\"]).count().reset_index() \n",
    "    # We make a sunburst chart : \n",
    "    # And another one with the same kind of information : a treemap:\n",
    "    treemap = px.treemap(sequence_df, path=[px.Constant(\"\"), 'nter', 'flag', 'cter', 'Date'], values='No of sequences',\n",
    "                    color='nter',color_discrete_map = NTER_CMAP)\n",
    "    treemap.update_traces(root_color=\"lightgrey\")\n",
    "    treemap.update_layout(title=\"No of Calcyanin.\", margin = dict(t=50, l=25, r=25, b=25))\n",
    "    return treemap\n",
    "\n",
    "#make_calcyanin_treemap(cnx)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "c58158b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_data(cnx):\n",
    "    # We convert the \"strain\" dataframe into a dictionnary - later  we will inject those datas into the report file using jinja2.\n",
    "    DATAS = {}\n",
    "\n",
    "    genomes_df = pd.read_sql_query(\"\"\"\n",
    "        SELECT * FROM harley as h  JOIN \n",
    "        genomes as g on g.Accession = h.Accession LEFT JOIN \n",
    "        checkm as c on g.Accession=c.`Bin Id` LEFT JOIN\n",
    "        gtdbtk as t on g.Accession=t.`user_genome`\n",
    "        \"\"\",cnx)\n",
    "    genomes_df = genomes_df.fillna(\"NA\")\n",
    "    genomes_df = genomes_df.loc[:,~genomes_df.columns.duplicated()]\n",
    "    genomes_df = genomes_df[['Accession', 'Assembly name','Date', 'Submitter',\n",
    "       'Submission date', 'Isolate', 'TaxID', 'Organism', 'Biosample',\n",
    "       'Isolation source', 'Environment (biome)', 'Geographic location',\n",
    "       'Culture collection', 'Collection date', 'Sample type',\n",
    "       'Completeness', 'Contamination', 'Strain heterogeneity',\n",
    "       'Genome size (bp)', '# scaffolds', '# contigs',\n",
    "       'N50 (scaffolds)', 'N50 (contigs)','# predicted genes',\n",
    "       'classification', 'fastani_reference','fastani_ani']]\n",
    "    #keys are the name of the organism, value are dictionnaries (gid : datas)\n",
    "    for org , sdf in genomes_df.groupby(\"Organism\"):\n",
    "        sdf.index = sdf.Accession\n",
    "        DATAS[org]=sdf.T.to_dict()\n",
    "        for acc in sdf.index:\n",
    "            seqs = pd.read_sql_query(\"\"\"\n",
    "                SELECT * FROM summary as s JOIN \n",
    "                ccya as c on c.sequence_id=s.sequence_accession WHERE\n",
    "                s.sequence_src=\"{}\"\n",
    "                \"\"\".format(acc),cnx,index_col=\"sequence_id\")\n",
    "            seqs = seqs.fillna(\"NA\")\n",
    "            seqs = seqs.T.to_dict()\n",
    "            DATAS[org][acc][\"sequences\"] = seqs\n",
    "            for seq in seqs.keys():\n",
    "                features = pd.read_sql_query(\"\"\"\n",
    "                    SELECT * FROM features as f WHERE\n",
    "                        f.sequence_id=\"{}\"\n",
    "                \"\"\".format(seq),cnx,index_col=\"feature_id\")\n",
    "                features = features.fillna(\"NA\")\n",
    "                DATAS[org][acc][\"sequences\"][seq][\"features\"] = features.T.to_dict()\n",
    "\n",
    "    # we convert the dictionnary into a json object for jinja2 injection.\n",
    "    json_object = json.dumps(DATAS, indent = 4)     \n",
    "    return json_object\n",
    "\n",
    "#make_data(cnx)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "2fdd8472",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# And we fill the template \n",
    "templatedir = \"../../../../../WORKFLOWS/harley/report\"\n",
    "with open(os.path.join(templatedir,'template.html')) as file_:\n",
    "    template = Template(file_.read())\n",
    "\n",
    "oms_plots = make_modorg_chart(cnx)\n",
    "    \n",
    "report = template.render(\n",
    "    dbname = \"TEST\",\n",
    "    datas  = make_data(cnx),\n",
    "    css= open( os.path.join(templatedir,\"template.css\" )).read(),\n",
    "    js = open( os.path.join(templatedir,\"template.js\"  )).read(),\n",
    "    \n",
    "    decision_tree = make_decision_tree_chart().to_html().split(\"<body>\")[1].split(\"</body>\")[0],\n",
    "    \n",
    "    sunburst = make_sunburst(cnx).to_html().split(\"<body>\")[1].split(\"</body>\")[0],# if sunburst else \"<p style='font-style: italic;'>No calcyanin detected - No sunburst :/ </p>\"  ,\n",
    "    \n",
    "    treemap = make_calcyanin_treemap(cnx).to_html().split(\"<body>\")[1].split(\"</body>\")[0],# if sunburst else \"<p style='font-style: italic;'>No calcyanin detected - No treemap :/ </p>\"  ,\n",
    "    \n",
    "    cobahma_oms = oms_plots[\"CoBaHMA-type\"].to_html().split(\"<body>\")[1].split(\"</body>\")[0] if \"CoBaHMA-type\" in oms_plots else \"<p style='font-style: italic;'>No data for this kind of N-ter</p>\"  ,\n",
    "    \n",
    "    x_oms = oms_plots[\"X-type\"].to_html().split(\"<body>\")[1].split(\"</body>\")[0] if \"X-type\" in oms_plots else \"<p style='font-style: italic;'>No data for this kind of N-ter</p>\"  ,\n",
    "    \n",
    "    y_oms = oms_plots[\"Y-type\"].to_html().split(\"<body>\")[1].split(\"</body>\")[0] if \"Y-type\" in oms_plots else \"<p style='font-style: italic;'>No data for this kind of N-ter</p>\"  ,\n",
    "    \n",
    "    z_oms = oms_plots[\"Z-type\"].to_html().split(\"<body>\")[1].split(\"</body>\")[0] if \"Z-type\" in oms_plots else \"<p style='font-style: italic;'>No data for this kind of N-ter</p>\"  ,\n",
    "    \n",
    "    unknown_oms = oms_plots[\"Unknown-type\"].to_html().split(\"<body>\")[1].split(\"</body>\")[0]  if \"Unknown-type\" in oms_plots else \"<p style='font-style: italic;'>No data for this kind of N-ter</p>\"  ,\n",
    "    \n",
    "    genome_over_time = make_genome_over_time_chart(cnx).to_html().split(\"<body>\")[1].split(\"</body>\")[0],\n",
    "    \n",
    "    sequence_over_time = make_sequence_over_time_chart(cnx).to_html().split(\"<body>\")[1].split(\"</body>\")[0],\n",
    "    \n",
    "    metrics_fig = make_genome_pie_chart(cnx).to_html(config= dict(displayModeBar = False)).split(\"<body>\")[1].split(\"</body>\")[0],\n",
    "    \n",
    "    #strain_total = len(genomes_metrics[\"Strain\"][\"strains\"]),    \n",
    "    \n",
    "    #cr_total =  len(genomes_metrics[\"GenBanK-RefSeq\"][\"genomes\"]) , \n",
    "    \n",
    "    #gb_total =  len(genomes_metrics[\"GenBanK\"][\"genomes\"]) , \n",
    "    \n",
    "    #rs_total =  len(genomes_metrics[\"RefSeq\"][\"genomes\"]) , \n",
    "    \n",
    "    #total = len(genomes_metrics[\"RefSeq\"][\"genomes\"]) + len(genomes_metrics[\"GenBanK\"][\"genomes\"]),\n",
    ")\n",
    "report\n",
    "\n",
    "def save(file,report):\n",
    "    dirname = os.path.abspath(os.path.dirname(file))\n",
    "    os.makedirs(dirname,exist_ok=True)\n",
    "    with open( file , 'w' ) as stream:\n",
    "        stream.write(report)\n",
    "    \n",
    "save(\"test.html\",report)\n",
    "\n",
    "os.system(\"open test.html\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "ade26a55",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "72"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "3*24"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "01647bb1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "82ae3943",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
